// @instantdb/react/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstantReact = exports.i = exports.Cursors = exports.InstantReactWeb = exports.init_experimental = exports.init = exports.lookup = exports.tx = exports.id = void 0;
const core_1 = require("@instantdb/core");
Object.defineProperty(exports, "id", { enumerable: true, get: function () { return core_1.id; } });
Object.defineProperty(exports, "tx", { enumerable: true, get: function () { return core_1.tx; } });
Object.defineProperty(exports, "lookup", { enumerable: true, get: function () { return core_1.lookup; } });
Object.defineProperty(exports, "i", { enumerable: true, get: function () { return core_1.i; } });
const InstantReact_1 = require("./InstantReact");
Object.defineProperty(exports, "InstantReact", { enumerable: true, get: function () { return InstantReact_1.InstantReact; } });
const InstantReactWeb_1 = require("./InstantReactWeb");
Object.defineProperty(exports, "InstantReactWeb", { enumerable: true, get: function () { return InstantReactWeb_1.InstantReactWeb; } });
const init_1 = require("./init");
Object.defineProperty(exports, "init", { enumerable: true, get: function () { return init_1.init; } });
Object.defineProperty(exports, "init_experimental", { enumerable: true, get: function () { return init_1.init_experimental; } });
const Cursors_1 = require("./Cursors");
Object.defineProperty(exports, "Cursors", { enumerable: true, get: function () { return Cursors_1.Cursors; } });
