// @instantdb/core/dist/schemaTypes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EntityDef = exports.InstantGraph = exports.DataAttrDef = exports.LinkAttrDef = void 0;
class LinkAttrDef {
    constructor(entityName, cardinality) {
        this.entityName = entityName;
        this.cardinality = cardinality;
    }
}
exports.LinkAttrDef = LinkAttrDef;
class DataAttrDef {
    constructor(valueType, required, config = { indexed: false, unique: false }) {
        this.valueType = valueType;
        this.required = required;
        this.config = config;
    }
    optional() {
        return new DataAttrDef(this.valueType, false);
    }
    unique() {
        return new DataAttrDef(this.valueType, this.required, Object.assign(Object.assign({}, this.config), { unique: true }));
    }
    indexed() {
        return new DataAttrDef(this.valueType, this.required, Object.assign(Object.assign({}, this.config), { indexed: true }));
    }
}
exports.DataAttrDef = DataAttrDef;
class InstantGraph {
    constructor(entities, links) {
        this.entities = entities;
        this.links = links;
    }
    withRoomSchema() {
        return new InstantGraph(this.entities, this.links);
    }
}
exports.InstantGraph = InstantGraph;
class EntityDef {
    constructor(attrs, links) {
        this.attrs = attrs;
        this.links = links;
    }
    asType() {
        return new EntityDef(this.attrs, this.links);
    }
}
exports.EntityDef = EntityDef;
